File: exercise/ts/day03/tests/santaWorkshopService.spec.ts
import {SantaWorkshopService} from "../src/santaWorkshopService";
import {Gift} from "../src/gift";
import { faker } from '@faker-js/faker';

describe('SantaWorkshopService', () => {
    let service: SantaWorkshopService;

    beforeEach(() => {
        service = new SantaWorkshopService();
    });

    const validWeight = () => faker.number.float({min: 0, max: 5, fractionDigits: 2});
    const invalidWeight = () => faker.number.float({min: 6, max: Number.MAX_SAFE_INTEGER, fractionDigits: 2})
    const prepareGiftFor = (weight: number) => {
        const giftName = faker.commerce.productName();
        const color = faker.color.human()
        const material = faker.commerce.productMaterial();

        return service.prepareGift(giftName, weight, color, material);
    }

    it('should prepare a gift with valid parameters', () => {
        const gift = prepareGiftFor(validWeight())
        expect(gift).toBeInstanceOf(Gift);
    });

    it('should throw an error if gift is too heavy', () => {
        expect(() => prepareGiftFor(invalidWeight())).toThrow('Gift is too heavy for Santa\'s sleigh');
    });

    it('should add an attribute to a gift', () => {
        const gift = prepareGiftFor(validWeight())
        gift.addAttribute('recommendedAge', '3');

        expect(gift.getRecommendedAge()).toBe(3);
    });
});

File: exercise/scala/day03/src/test/scala/SantaWorkshopServiceSpec.scala
import org.scalatest.funspec.AnyFunSpec
import org.scalatest.matchers.should.Matchers
import preparation.{Gift, SantaWorkshopService}

class SantaWorkshopServiceSpec extends AnyFunSpec with Matchers {
  describe("SantaWorkshopService") {
    val service = new SantaWorkshopService

    describe("prepareGift") {
      it("should prepare a gift with valid parameters") {
        val giftName = "Bitzee"
        val weight = 3.0
        val color = "Purple"
        val material = "Plastic"

        service.prepareGift(giftName, weight, color, material) should not be null
      }

      it("should retrieve an attribute to a gift") {
        val giftName = "Furby"
        val weight = 1.0
        val color = "Multi"
        val material = "Cotton"

        val gift = new Gift(giftName, weight, color, material)
        gift.addAttribute("recommendedAge", "3")

        gift.getRecommendedAge shouldBe 3
      }

      it("should fail if gift is too heavy") {
        val giftName = "Dog-E"
        val weight = 6.0
        val color = "White"
        val material = "Metal"

        val exception = intercept[IllegalArgumentException] {
          service.prepareGift(giftName, weight, color, material)
        }
        exception.getMessage shouldBe "Gift is too heavy for Santa's sleigh"
      }
    }
  }
}